From 363c33535b3ec6c08a7bb760f64c0f0177f2cc0b Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Mon, 3 Aug 2020 11:14:01 -0700
Subject: [PATCH 1506/1851] drm: xlnx: mixer: Fix async update kernel warning

commit 394d06094db3f261067a738639233b906f5cb066 from
https://github.com/Xilinx/linux-xlnx.git

This patch fixes the below warning when Video Mixer is connected to HDMI
Tx.

[   43.728875  ] ------------[ cut here   ]------------
[   43.733620  ] WARNING: CPU: 3 PID: 1047 at
drivers/gpu/drm/drm_atomic_helper.c:1744
drm_atomic_helper_async_commit+0xec/0x148
[   43.744978  ] Modules linked in: xilinx_hdmi_tx(O) xilinx_vphy(O)
dp159(O)
[   43.751831  ] CPU: 3 PID: 1047 Comm: videotestsrc0:s Tainted: G
O      5.4.0-00137-gf6cfcafe827b-dirty #94
[   43.762392  ] Hardware name: ZynqMP ZCU102 Rev1.0 (DT)
[   43.767459  ] pstate: 80000005 (Nzcv daif -PAN -UAO)
[   43.772342  ] pc : drm_atomic_helper_async_commit+0xec/0x148
[   43.777938  ] lr : drm_atomic_helper_async_commit+0x68/0x148
[   43.783536  ] sp : ffff8000113039d0
[   43.786909  ] x29: ffff8000113039d0 x28: ffff00086f837d00
[   43.792326  ] x27: 0000000000000000 x26: 0000000000000000
[   43.797747  ] x25: 0000000000000500 x24: ffff00086f837d00
[   43.803173  ] x23: ffff000879784400 x22: ffff00086f813e00
[   43.808598  ] x21: 0000000000000005 x20: ffff00087ae9d8e0
[   43.814024  ] x19: ffff00086f837700 x18: 0000000000000000
[   43.819449  ] x17: 0000000000000000 x16: 0000000000000000
[   43.824875  ] x15: 0000000000000000 x14: ffff0008716e4200
[   43.830300  ] x13: 0000000000000000 x12: 0000000000000000
[   43.835726  ] x11: 0000000000000000 x10: ffff00087976fc80
[   43.841152  ] x9 : 0000000000000050 x8 : 0000000000000007
[   43.846577  ] x7 : 0000000000000000 x6 : 0000000000000000
[   43.852002  ] x5 : 0000000000000000 x4 : 0000000000000000
[   43.857428  ] x3 : ffff00086f837600 x2 : 0000000000000000
[   43.862853  ] x1 : 0000000000000000 x0 : ffff00086f837d00
[   43.868279  ] Call trace:
[   43.870779  ]  drm_atomic_helper_async_commit+0xec/0x148
[   43.876021  ]  drm_atomic_helper_commit+0xd8/0x140
[   43.880733  ]  drm_atomic_commit+0x48/0x58
[   43.884737  ]  xlnx_mix_disp_plane_atomic_update_plane+0x114/0x168
[   43.890867  ]  __setplane_atomic+0xf0/0x150
[   43.894963  ]  drm_mode_setplane+0xf8/0x2b8
[   43.899052  ]  drm_ioctl_kernel+0xb8/0x108
[   43.903051  ]  drm_ioctl+0x200/0x410
[   43.906523  ]  do_vfs_ioctl+0x878/0xa20
[   43.910260  ]  ksys_ioctl+0x44/0x90
[   43.913638  ]  __arm64_sys_ioctl+0x1c/0x28
[   43.917647  ]  el0_svc_common.constprop.0+0x68/0x160
[   43.922534  ]  el0_svc_handler+0x6c/0x88
[   43.926356  ]  el0_svc+0x8/0xc
[   43.929295  ] ---[ end trace d8e4931ce03e45b5   ]---

This occurs only for the first time when modetest is run followed
by gst-launch. Subsequent runs don't see this warning.

1. modetest -D 80070000.v_mix -s 44@42:1280x720@BG24
2. gst-launch-1.0 videotestsrc ! video/x-raw, width=1280, height=720,
format=NV12 ! kmssink bus-id="80070000.v_mix" plane-id=39

The drm_atomic_set_fb_for_plane() doesn't unset or swap new_state->fb, so
drm_atomic_helper_async_commit() doesn't see the old fb on plane state
after async update. Hence use swap() to update new_state->fb.

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/xlnx_mixer.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/xlnx/xlnx_mixer.c b/drivers/gpu/drm/xlnx/xlnx_mixer.c
index dc4d6d76cb48..826217b74664 100644
--- a/drivers/gpu/drm/xlnx/xlnx_mixer.c
+++ b/drivers/gpu/drm/xlnx/xlnx_mixer.c
@@ -2031,7 +2031,7 @@ xlnx_mix_plane_atomic_async_update(struct drm_plane *plane,
 		drm_atomic_get_old_plane_state(new_state->state, plane);
 
 	/* Update the current state with new configurations */
-	drm_atomic_set_fb_for_plane(plane->state, new_state->fb);
+	swap(plane->state->fb, new_state->fb);
 	plane->state->crtc = new_state->crtc;
 	plane->state->crtc_x = new_state->crtc_x;
 	plane->state->crtc_y = new_state->crtc_y;
-- 
2.31.1

